x-pack/plugins/fleet/server/services/check_allowed_packages.ts

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { get } from 'lodash';
+
+import { FleetUnauthorizedError } from '../errors';
+
+export function filterAllowedPackages<T>(
+  packages: T[],
+  keyPath: string,
+  allowedPackages?: string[]
+): T[] {
+  // allowedPackages is undefined when user has privileges from fleet
+  if (!packages.length || !allowedPackages) {
+    return packages;
+  }
+
+  if (!allowedPackages.length) {
+    return [];
+  }
+
+  const allowedPackagedSet = new Set(allowedPackages);
+  return packages.filter((pkg) => {
+    const pkgName = get(pkg, keyPath);
+    return allowedPackagedSet.has(pkgName);
+  });
+}
+
+export function checkAllowedPackages<T>(
+  packages: T[],
+  keyPath: string,
+  allowedPackages?: string[]
+) {
+  // allowedPackages is undefined when user has privileges from fleet
+  if (!packages.length || !allowedPackages) {
+    return;
+  }
+
+  if (!allowedPackages.length) {
+    throw new FleetUnauthorizedError(
+      'Authorization denied due to lack of integration package privileges'
+    );
+  }
+
+  const allowedPackagedSet = new Set(allowedPackages);
+  const allowedPackagesStr = allowedPackages.join(', ');
+  const restrictedPackages = packages.reduce((acc, pkg) => {
+    const pkgName = get(pkg, keyPath, undefined);
+    if (!pkgName) {
+      throw new FleetUnauthorizedError(
+        `Authorization denied. Allowed package(s): ${allowedPackagesStr}.`
+      );
+    }
+
+    const isRestricted = !allowedPackagedSet.has(pkgName);
+    if (!isRestricted) return acc;
+
+    return [...acc, pkgName];
+  }, [] as string[]);
```

Comment:
I was thinking this would make for a better UX so users can request all permissions from admin in one swoop but I can see how it could be a security concern and it being potentially large is certainly another issue. Easy enough to change 👌.